File: backend/app/schemas/panoramica.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, model_validator


class PanoramicaVoceCreate(BaseModel):
    modulo: str
    sezione_slug: str
    etichetta: str
    campo: str | None = None
    record_id: uuid.UUID | None = None

    @model_validator(mode="after")
    def _campo_xor_record_id(self) -> "PanoramicaVoceCreate":
        if (self.campo is None) == (self.record_id is None):
            raise ValueError("Indicare esattamente uno tra 'campo' (sezione singleton) e 'record_id' (sezione a elenco)")
        return self


class PanoramicaVoceRead(PanoramicaVoceCreate):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    azienda_id: uuid.UUID
    ordine: int
    created_at: datetime
    updated_at: datetime


class PanoramicaOrdineUpdate(BaseModel):
    modulo: str
    ids: list[uuid.UUID]
    """Elenco degli id delle voci nel nuovo ordine desiderato: la posizione
    nell'elenco diventa il nuovo valore di `ordine` di ciascuna voce."""

File: frontend/lib/types/panoramica.ts
// Tipi TypeScript allineati ai contratti Pydantic esposti da
// backend/app/schemas/panoramica.py.

export interface PanoramicaVoce {
  id: string;
  azienda_id: string;
  modulo: string;
  sezione_slug: string;
  // Esattamente uno dei due è valorizzato: campo per le sezioni singleton,
  // record_id per un intero record di una sezione a elenco.
  campo: string | null;
  record_id: string | null;
  etichetta: string;
  ordine: number;
  created_at: string;
  updated_at: string;
}
